aten/src/ATen/native/cuda/IndexKernel.cu

```diff
@@ -103,6 +107,117 @@ static void index_put_kernel(TensorIterator& iter, IntArrayRef index_size, IntAr
   });
 }
 
+template<typename scalar_t, typename mask_t>
+void masked_select_out_cuda_kernel(
+  scalar_t* result_ptr,
+  scalar_t* self_ptr,
+  mask_t* mask_ptr,
+  int64_t* mask_inclusive_scan_ptr,
+  int64_t num_input_elements
+) {
+  legacy::launch_kernel<launch_size_nd, launch_bound2>(
```

Comment:
Why there is a `legacy::launch_kernel`? Is it possible to refactor it to use `gpu_kernel`? Legacy kernels are going to be removed soon: https://github.com/pytorch/pytorch/pull/34749

Reply:
Thanks for pointing that out, I didn't know about gpu_kernel before. I'm not 100% sure how best to use it in this case though. gpu_kernel and TensorIterator expect to be directly involved with an output tensor. But the location in my output tensor whether and where each thread of the kernel needs to place its output is dependent on input data, `_mask` and `mask_inclusive_scan`. So I don't think that I can use `TensorIterator::add_output()` or the lambda's return statement to make outputs go to the right place, unless there's some method of handling that that I don't know about.

What I figured I could do is keep using my pointer-style accesses for the output tensor. The input tensors can be accessed in the TensorIterator style, with `add_input()`. But it seems like TensorIterator requires an output tensor, so I just put a dummy tensor in there and return 0 from the lambda. But that will be extra work--the dummy tensor is going to be resized and the gpu_kernel call is going to fill it with 0. Maybe there's a setting in TensorIterator or gpu_kernel that can let me turn off outputs?

Reply:
I see the point, the assumption that GPU loops have 1 and only 1 output is baked deep, so it's hard to change this behavior. But instead of using the legacy kernel, does it make sense to write your own simple kernel instead? It could be something like:
```CUDA
template<typename scalar_t, typename mask_t>
__global__ masked_select_kernel(
      scalar_t* result_ptr,
      scalar_t* self_ptr,
      mask_t* mask_ptr,
      int64_t* mask_inclusive_scan_ptr,
      int64_t num_input_elements
) {
    mask_t mask[thread_work_size];
    int64_t base_index = block_work_size * blockIdx.x;
    int remaining = std::min<int64_t>(num_input_elements - base_index, block_work_size);

    // load data into registers
    int thread_idx = threadIdx.x;
    #pragma unroll
    for(int i = 0; i < thread_work_size; i++) {
        if (thread_idx >= remaining) {
          break;
        }
        int64_t input_idx = thread_idx + base_index;
        mask[i] = mask_ptr[input_idx];
        thread_idx += num_threads;
    }

    // compute and store
    int thread_idx = threadIdx.x;
    #pragma unroll
    for(int i = 0; i < thread_work_size; i++) {
        if (thread_idx >= remaining) {
            break;
        }
        if (mask[i]) {
            int64_t input_idx = thread_idx + base_index;
            int64_t result_idx = mask_inclusive_scan_ptr[input_idx] - 1;
            result_ptr[result_idx] = self_ptr[input_idx];
        }
        thread_idx += num_threads;
    }
}
```
And you can launch the kernel with
```
auto stream = at::cuda::getCurrentCUDAStream();
int64_t grid = (N + block_work_size - 1) / block_work_size;
masked_select_kernel<<<grid, num_threads, 0, stream>>>(.....);
```
Where `block_work_size`, `num_threads`, and `thread_work_size` are all consts defined in `Loops.cuh`.

Reply:
Ok I can do that.

Reply:
Thanks for the help, @zasdfgbnm! Your code gives a nice performance improvement. I'll post new measurements shortly.

Also, I tried combining your two loops but it gave slower performance. So I'm assuming your strategy with two loops was to limit branch divergence by executing conditionals closer to the end of the kernel, is that correct? I'll need to remember that trick.

Reply:
This code allows ILP (loads can be issues without waiting for the result, and then stores). If you combine 2 loops, compiler is not smart enough to unroll and do that.

Reply:
Oh that makes sense.